File: src/lib/scala/Slate.Common/src/main/scala/slate/common/results/ResultTimed.scala
/**
  * <slate_header>
  * author: Kishore Reddy
  * url: https://github.com/kishorereddy/scala-slate
  * copyright: 2015 Kishore Reddy
  * license: https://github.com/kishorereddy/scala-slate/blob/master/LICENSE.md
  * desc: a scala micro-framework
  * usage: Please refer to license on github for more info.
  * </slate_header>
  */

package slate.common.results

import java.time.Duration

import slate.common._
import slate.common.info.MemUsage

case class ResultTimed[+T](
                        desc        : String,
                        start       : DateTime,
                        end         : DateTime,
                        duration    : Duration,
                        result      : Result[T],
                        memory      : Option[MemUsage],
                        avg         : Option[Duration] = None
                       )
{
  val dur = start.durationFrom(end)

  /**
   *
   * @param start
   * @param end
   * @param duration
   * @param mem
   * @param data
   */
  def this(start:DateTime, end:DateTime, duration:Duration, data:Result[T], mem:Option[MemUsage], avg:Option[Duration])
  {
    this("", start, end, duration, data, mem, avg)
  }


  def print():Unit = {
    println("desc           : " + desc               )
    println("start          : " + start              )
    println("end            : " + end                )
    println("duration.secs  : " + duration.getSeconds)
    println("duration.nanos : " + duration.getNano   )

    memory.map(mem => {
      println("memory.used    : " + mem.used)
      println("memory.free    : " + mem.free)
      println("memory.total   : " + mem.total)
      println("memory.max     : " + mem.max)
      Unit
    })
    println()
  }


  def toJson():String = {
    "not implemented"
  }
}


object ResultTimed {

  def build[T](started:DateTime, result:Result[T] ):ResultTimed[T] = {
    val ended = DateTime.now()
    val duration = started.durationFrom(ended)
    new ResultTimed[T]("", started, ended, duration, result, None, None)
  }
}

File: src/lib/kotlin/slatekit-common/src/main/kotlin/slatekit/common/results/ResultTimed.kt
/**
 * <slate_header>
 * url: www.slatekit.com
 * git: www.github.com/code-helix/slatekit
 * org: www.codehelix.co
 * author: Kishore Reddy
 * copyright: 2016 CodeHelix Solutions Inc.
 * license: refer to website and/or github
 * about: A tool-kit, utility library and server-backend
 * mantra: Simplicity above all else
 * </slate_header>
 */

package slatekit.common.results

import slatekit.common.DateTime
import slatekit.common.ResultEx
import slatekit.common.info.Memory
import java.time.Duration

data class ResultTimed<T>(
        val desc: String,
        val start: DateTime,
        val end: DateTime,
        val duration: Duration,
        val result: ResultEx<T>,
        val memory: Memory?,
        val avg: Long?
) {
    val dur = start.durationFrom(end)

    /**
     *
     * @param start
     * @param end
     * @param duration
     * @param mem
     * @param data
     */
    constructor(start: DateTime, end: DateTime, duration: Duration, data: ResultEx<T>, mem: Memory?, avg: Long?)
            :
            this("", start, end, duration, data, mem, avg)


    fun print(): Unit {
        println("desc           : " + desc)
        println("start          : " + start)
        println("end            : " + end)
        println("duration.secs  : " + duration.seconds)
        println("duration.nanos : " + duration.nano)

        if (memory != null) {
            println("memory.used    : " + memory.used)
            println("memory.free    : " + memory.free)
            println("memory.total   : " + memory.total)
            println("memory.max     : " + memory.max)
            Unit
        }
        println()
    }


    companion object {

        @JvmStatic
        fun <T> build(started: DateTime, result: ResultEx<T>): ResultTimed<T> {
            val ended = DateTime.now()
            val duration = started.durationFrom(ended)
            return ResultTimed<T>("", started, ended, duration, result, null, null)
        }
    }

}